### PR TITLE
Properly support user-provided norm.

### LINKED ISSUE #2381: Plotting norm is superseded by xarray vmin and vmax
#### Code Sample

```python
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

# prepare array with values between 1e0 and 1e3
da = xr.DataArray(np.logspace(0, 3, 24).reshape(4, 6), dims=('x', 'y'))

# but we want to log-normalize between 1e0 and 1e2
norm = mcolors.LogNorm(1e1, 1e2)

# plot with log norm
da.plot.imshow(norm=norm)  # log norm is applied but bounds are replaced
#da.plot.imshow(norm=norm, vmin=norm.vmin, vmax=norm.vmax)  # works, but redundant

# show results
plt.show()

```
#### Problem description

When passing a norm argument to imshow etc, the norm is forwarded to matplotlib but its data bounds are replaced with those calculated by xarray.

#### Expected Output

I think that the norm's ``vmin`` and ``vmax`` should take precedence over xarray-generated ``vmin`` and ``vmax``.

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.15.final.0
python-bits: 64
OS: Linux
OS-release: 4.14.65-1-MANJARO
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: fr_CH.utf8
LOCALE: None.None

xarray: 0.10.8
pandas: 0.23.4
numpy: 1.15.0
scipy: 1.1.0
netCDF4: 1.4.1
h5netcdf: None
h5py: None
Nio: None
zarr: None
bottleneck: None
cyordereddict: None
dask: 0.18.2
distributed: None
matplotlib: 2.2.3
cartopy: 0.15.1
seaborn: None
setuptools: 40.0.0
pip: 18.0
conda: None
pytest: None
IPython: 5.8.0
sphinx: 1.7.6
</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #2381  (remove if there is no corresponding issue, which should only be the case for minor changes)
 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

### CODE REVIEW

File: xarray/tests/test_plot.py [hunk 1]
```diff
@@ -641,10 +641,10 @@ def test_build_discrete_cmap(self):
 
     @pytest.mark.slow
     def test_discrete_colormap_list_of_levels(self):
-        for extend, levels in [('max', [-1, 2, 4, 8, 10]), ('both',
-                                                            [2, 5, 10, 11]),
-                               ('neither', [0, 5, 10, 15]), ('min',
-                                                             [2, 5, 10, 15])]:
+        for extend, levels in [('max', [-1, 2, 4, 8, 10]),
```
Comment (dcherian) [inline xarray/tests/test_plot.py]: only whitespace

File: xarray/tests/test_plot.py [hunk 2]
```diff
@@ -658,10 +658,10 @@ def test_discrete_colormap_list_of_levels(self):
 
     @pytest.mark.slow
     def test_discrete_colormap_int_levels(self):
-        for extend, levels, vmin, vmax in [('neither', 7, None,
-                                            None), ('neither', 7, None, 20),
-                                           ('both', 7, 4, 8), ('min', 10, 4,
-                                                               15)]:
+        for extend, levels, vmin, vmax in [('neither', 7, None, None),
```
Comment (dcherian) [inline xarray/tests/test_plot.py]: only whitespace

File: xarray/tests/test_plot.py [hunk 3]
```diff
@@ -687,6 +687,13 @@ def test_discrete_colormap_list_levels_and_vmin_or_vmax(self):
         assert primitive.norm.vmax == max(levels)
         assert primitive.norm.vmin == min(levels)
 
+    def test_colormap_pass_norm(self):
+        norm = mpl.colors.LogNorm(0.1, 1e1)
+        primitive = self.darray.plot(norm=norm, vmin=2, vmax=5)
```
Comment (shoyer) [inline xarray/tests/test_plot.py]: Should we raise an error in this case instead?

**Comment by dcherian** [comment]
This turns out to be more complicated because you can create a `norm` object that doesn't have `vmin, vmax` specified. I'll send in a new PR.

**Pull request closed without merging**